### PR TITLE
Add document global events (database:*)

### PR DESCRIPTION
**Pull request by jenow**
## What does this PR do ?

:warning: This PR is the same as https://github.com/kuzzleio/kuzzle/pull/1324.
It's a duplicate because it has been merge and it was a mistake so we reverted everythihng :warning: 

- add document global event

### How should this be manually tested?

- create a plugin pipes `database:beforeWrite`
- call action `document:create`
- see your pipe called with documents

### Other changes

/

### Boyscout

- fix a possible error in a functional test
```javascript
else if(response.errors === null && withErrors) { // response.error not errors
  callback('Should get partial error');
   return false;
}
```

### CODE REVIEW

File: lib/config/documentEventAliases.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+module.exports = {
+  namespace: 'database',
```
Comment (Aschen) [inline lib/config/documentEventAliases.js]: We agree on `generic:document:*` for theses events (ping @thomasarbona )

File: features/step_definitions/deleteDocument.js [hunk 2]
```diff
@@ -61,11 +61,12 @@ Then(/^I remove the documents '([^']+)'( and get partial errors)?$/, function (d
 
   this.api.mDelete({ids: documents})
     .then(response => {
+      
       if (response.error !== null && !withErrors) {
```
Comment (Yoann-Abbes) [inline features/step_definitions/deleteDocument.js]:
```suggestion
      if (response.error && !withErrors) {
```

File: features/step_definitions/deleteDocument.js [hunk 3]
```diff
@@ -61,11 +61,12 @@ Then(/^I remove the documents '([^']+)'( and get partial errors)?$/, function (d
...
         callback(response.error.message);
         return false;
       }
-      else if(response.errors === null && withErrors) {
+      else if(response.error === null && withErrors) {
```
Comment (Yoann-Abbes) [inline features/step_definitions/deleteDocument.js]:
```suggestion
      else if(!response.error && withErrors) {
```

File: lib/api/controllers/funnelController.js [hunk 4]
```diff
@@ -398,6 +414,29 @@ class FunnelController {
...
+    if (!mustTrigger) {
+      return Bluebird.resolve(request);
+    }
+
+    const alias = this.documentEventAliases.mirrorList[action];
+    const event = `${this.documentEventAliases.namespace}:${prefix}${capitalizeFirstLetter(alias)}`;
+    const extractor = new DocumentExtractor(request);
+    return this.kuzzle
+      .pipe(event, extractor.extract(), request)
+      .then((documents) => extractor.insert(documents));
```
Comment (kuzzle) [inline lib/api/controllers/funnelController.js]: MINOR Remove parentheses around the parameter of this arrow function. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS3524)

**Comment by kuzzle** [comment]
SonarQube analysis reported 2 issues
* ![MINOR][MINOR] 1 minor
* ![INFO][INFO] 1 info

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![INFO][INFO] [pluginsManager.js#L759](https://github.com/kuzzleio/kuzzle/blob/54814b96dd613c1ad1d960739ed49b41fa32db41/lib/api/core/plugins/pluginsManager.js#L759): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by kuzzle** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![INFO][INFO] [pluginsManager.js#L759](https://github.com/kuzzleio/kuzzle/blob/5eb3f18eb930272d3228a492e01eab6404a0d077/lib/api/core/plugins/pluginsManager.js#L759): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

**Comment by kuzzle** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![INFO][INFO] [pluginsManager.js#L758](https://github.com/kuzzleio/kuzzle/blob/fc72f033cbc74d04d875d271828b4c5d62b1c5d8/lib/api/core/plugins/pluginsManager.js#L758): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

**Comment by kuzzle** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![INFO][INFO] [pluginsManager.js#L758](https://github.com/kuzzleio/kuzzle/blob/72aa18f9538da91a71e76f12dbc48319871c411e/lib/api/core/plugins/pluginsManager.js#L758): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

File: lib/api/core/plugins/pluginsManager.js [hunk 5]
```diff
@@ -418,7 +421,7 @@ class PluginsManager {
         }, timeoutDelay);
       }
 
-      const cb = (err, pipeResponse) => {
+      const cb = (err, pipeResponse, ...info2) => {
```
Comment (alexandrebouthinon) [inline lib/api/core/plugins/pluginsManager.js]: You should find a better name than `info2`

File: lib/api/core/plugins/pluginsManager.js [hunk 6]
```diff
@@ -427,18 +430,14 @@ class PluginsManager {
         }
 
         if (!timedOut) {
-          callback(err, pipeResponse);
+          callback(err, pipeResponse, ...info2);
```
Comment (alexandrebouthinon) [inline lib/api/core/plugins/pluginsManager.js]: You should find a better name than info2

**Comment by kuzzle** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![INFO][INFO] [pluginsManager.js#L758](https://github.com/kuzzleio/kuzzle/blob/95a48bd21780778927033adfd864f586337a3b03/lib/api/core/plugins/pluginsManager.js#L758): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

File: lib/api/core/plugins/pluginsManager.js (hunk 5 above)
Comment (thomasarbona) [reply lib/api/core/plugins/pluginsManager.js]: Like what

**Pull request merged**